base/strings/substring.jl

```diff
@@ -221,7 +221,7 @@ end
     return n
 end
 
-function string(a::Union{Char, String, SubString{String}, Symbol}...)
+@assume_effects :foldable :removable function string(a::Union{Char, String, SubString{String}, Symbol}...)
```

Comment:
Can we go with `:foldable` and have the compiler prove nothrow based on the success of constprop? I don't usually like to annotate `:nothrow` unless it's required since it's easy to screw up.

Reply:
the nothrow annotation is a lot stronger since it allows the compiler to remove code even if it doesn't know the const-ness of the input.

Reply:
I know, that's why it's scarier. It's always good to be economical with nothrow annotations.